src/transformStateWithClones.js

```
@@ -7,7 +7,30 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const history = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      currentState = { ...currentState, ...action.extraData };
+    }
+
+    if (action.type === 'removeProperties') {
+      currentState = { ...currentState };
+
+      for (const key of action.keysToRemove) {
+        delete currentState[key];
+      }
+    }
+
+    if (action.type === 'clear') {
+      currentState = {};
+    }
+
+    history.push(currentState);
+  }
```

There is no default/error handling for unknown action types. This violates the checklist item: "3. [CODE STYLE]: switch/case should always have default case for error handling." Without a default the function will silently ignore unknown actions and still `history.push(currentState)` (line 30), which can result in pushing the previous state instance again instead of treating the action as invalid. Add a `default` that throws an error (e.g., `throw new Error('Unknown action type: ' + action.type)`).